ScalarStages Interpreter

cryogenian opened the pull request:
[ch4832] [ch4833] [ch4834] [ch4835] [ch4836] [ch4837] [ch4838] 

All stages could be interpreted on 3.4, pivot and mask (and so cartesian) can't be interpreted on mongo older than 3.2, because of `$type`. We can implement `typeExpr` using implicit ordering if `3.2` version is needed, although both MS and Amazon dbs have very restricted aggregation framework version.

djspiewak commented:
Let's publish this as a snapshot so we can get QA looking at it while the PR is in review. Great stuff!

djspiewak commented:
Published as `3.0.12-aa929d1`

cryogenian commented on datasource/src/main/scala/quasar/physical/mongo/Interpreter.scala:
```diff
@@ -0,0 +1,90 @@
+      case List() =>
+        interpretation
+      case hd :: tail => refineStep(key, hd) match {
+        case None => interpretation
```
If it's impossible to push one operator, all subsequent are uninterpreted in the ds and intpereted in mimir

cryogenian commented on datasource/src/main/scala/quasar/physical/mongo/Mongo.scala:
```diff
@@ -143,20 +148,40 @@ class Mongo[F[_]: MonadResourceErr : ConcurrentEffect] private[mongo](
+        val newStages = ScalarStages(IdStatus.ExcludeId, result.stages)
+        val parsed = aggregated map  (x => (newStages, x map interpreter.mapper))
+        // versioning last stand
+        ConcurrentEffect[F].handleErrorWith(parsed)(e => fallback)
```
:+1:

cryogenian commented on datasource/src/main/scala/quasar/physical/mongo/Mongo.scala:
```diff
@@ -143,20 +148,40 @@ class Mongo[F[_]: MonadResourceErr : ConcurrentEffect] private[mongo](
+      else {
+        val aggregated = aggregate(collection, result.aggregators)
+        val newStages = ScalarStages(IdStatus.ExcludeId, result.stages)
+        val parsed = aggregated map  (x => (newStages, x map interpreter.mapper))
```
:+1:

cryogenian commented on datasource/src/main/scala/quasar/physical/mongo/interpreter/Pivot.scala:
```diff
@@ -0,0 +1,90 @@
+      vectorType: ColumnType.Vector)
+      : Option[List[Aggregator]] = {
+
+    if (version < Version(3, 4, 0)) None
```
Right, thank you! 🙂

cryogenian commented on datasource/src/main/scala/quasar/physical/mongo/interpreter/Project.scala:
```diff
@@ -0,0 +1,58 @@
+
+    def filterGuard(prj: E.Projection): Boolean = {
+      val hasNotIx: Boolean = prj.steps.toList.forall(E.isField)
+      version > (if (hasNotIx) Version(0, 0, 0) else Version(3, 2, 0)) // Not sure if $let works for < 2.0
```
:+1:

cryogenian commented:
Yeah, I was going to make separate prs with every `ParseInstruction` (in particular omitting `Mask` and `Cartesian`) initially, but during working on `Mask`, parse instructions became `ScalarStages` and everything was simplified a lot :D, also I could bring whole tests at once. 

I'm going to take a look how complicated would be migrating to matryoshka, and use it if it could be updated in short time. Although I'm not going to change `interpretation` domain logic (I mean aggregator pipeline generated should be the same), so, the logic should be reviewable even during changing.

rintcius commented:
@cryogenian Not sure if you read it like that, but to be clear I didn't mean to look into matryoshka now

cryogenian commented:
Per discussion with @rintcius I've addressed only those parts that is need to be fixed in this pr. Also I added this

https://app.clubhouse.io/data/story/5328/consider-using-pattern-functor-matryoshka-in-mongo-connector

djspiewak commented:
@rintcius Official mongo support policy:

- 3.4.x (for latest value of *x*)
- 3.6.x
- M.m.x for the latest *two* values of *M* and *m*

For example, right now the two latest mongo lines are 4.0.x and 3.6.x. So we support 4.0.x, 3.6.x, and 3.4.x. When the next major line comes out, we will support 4.2.x, 4.0.x, 3.6.x, and 3.4.x. When the next release *after that* comes out, we will support 4.4.x, 4.2.x, 3.6.x, and 3.4.x.

cryogenian commented:
3.0.12-a6baf1c version is released compatible with 116.1.0

cryogenian commented:
[ch5391]

rintcius commented:
I'm 👍 on the part that I reviewed and do https://app.clubhouse.io/data/story/5394/mongo-pushdown-mechanics-polishing separate from this PR.

cryogenian commented:
@wemrysi @jsantos17 Please, review `interpreter` package

jsantos17 commented on datasource/src/main/scala/quasar/physical/mongo/Interpreter.scala:
```diff
@@ -0,0 +1,90 @@
+    case ScalarStage.Cartesian(cartouches) => Cartesian(key, version, cartouches, this)
+  }
+
+  private def initialAggregators(idStatus: IdStatus): Aggregator = idStatus match {
```
I'm not sure I follow how `_id` relates to `IdStatus`. Would you mind explaining?

cryogenian replied:
`ScalarStages(idStatus, List())` should return 
+ `[id, value]` for `IncludeId`
+ `id` for `IdOnly` 
+ `value` for `ExcludeId` 

Where `id` is unique value for every row. That `id` can be any type and it's not ordered (if we believe comments in `ParseIntructionSpec`) Enumerating looks a bit weird since we have unique identifier for every row already.

cryogenian commented on datasource/src/main/scala/quasar/physical/mongo/MongoExpression.scala:
```diff
@@ -0,0 +1,201 @@
+  final case class Index(int: SInt) extends ProjectionStep
+
+  object ProjectionStep {
+    implicit val eqProjectionStep: Eq[ProjectionStep] = Eq.fromUniversalEquals
```
What is the difference? I thought that case classes `==` is value equality, not referential

djspiewak replied:
In this case there isn't a difference. But if you were to add something that in turn has a custom `Eq` to the algebra *later*, the compiler wouldn't catch it for you.

jsantos17 commented:
@alissapajer @cryogenian @djspiewak @wemrysi @rintcius @nicflores 

As a broader question for the backend team: why did we choose to do partial push down for Mongo? With the current QScript, I'm fairly confident we can do full QScript pushdown by restoring the old connnector and updating all the file system bits to match the new file system . I wager that would be less work than attempting to re-implement everything on top of `ScalarStages`. If we drop support for Mongo 3.2, nothing (but joins) should require mapReduce.

djspiewak commented:
@jsantos17 Because we're about to have much significantly complicated things that cannot be fully pushed down. Using the `ScalarStage` model gives us control over that, as well as considerably simplifying the task of the underlying connector.

jsantos17 commented:
>Because we're about to have much significantly complicated things that cannot be fully pushed down

Can you provide an example?

djspiewak commented:
@jsantos17 The UI is going to allow users to run arbitrary computations, including functions (such as `round`) which are not well implemented in Mongo. Additionally, we're going to enable row-level aggregations in the future. That alone I think kills full pushdown, but in addition to that, we're going to expand the set of reducers beyond the very restricted set that Mongo implements (Mimir implements several more which aren't exposed in qscript because of mongo).

Ultimately, doing full pushdown makes the assertion that we will never support anything in the UI or query algebra which mongo *doesn't* support in its aggregation framework. That's not an assertion I would want to make even if we didn't have immediate concrete plans to violate it.

cryogenian commented on datasource/src/main/scala/quasar/physical/mongo/Aggregator.scala:
```diff
@@ -0,0 +1,73 @@
+        "preserveNullAndEmptyArrays" -> true))
+  }
+
+  final case class Filter(obj: E.Object) extends Aggregator {
```
Sorry, could you elaborate? This thing is used as constructor for `$match` command only, nothing smarter 🙂

jsantos17 commented:
How about replacing the `QScript -> WorkflowBuilder` stage with a `ScalarStage -> WorkflowBuilder` stage instead of re-implementing everything from scratch? We can reused most of the old connector that way, while doing partial pushdown. WDYT @rintcius?

I don't mean to say this PR is wrong or anything like that, but I'd like to explore

cryogenian commented:
@jsantos17 

> As a note, type-filtering for arrays before `$unwind` is not necessary in general. Mongo will do the Right Thing™️ if you try to `$unwind` something that isn't an array. I believe the behavior can be tuned through options on `$unwind`. 

Right, one of the filters is unnecessary, but we need to check that projection is object before call `$objectToArray`

wemrysi commented on datasource/src/main/scala/quasar/physical/mongo/interpreter/Project.scala:
```diff
@@ -0,0 +1,58 @@
+        val projection =
+          E.key(uniqueKey) +/ fld
+        val match_ =
+          Aggregator.filter(E.Object(projection.toKey -> E.Object("$exists" -> E.Bool(true))))
```
What necessitates the `$exists` check vs just projecting?

cryogenian replied:
This is to `elide rows not containing path` :)

cryogenian replied:
Filtering `null` doesn't solve the problem btw, as that test emits `[]` then

wemrysi commented on datasource/src/test/scala/quasar/physical/mongo/StageInterpreterSpec.scala:
```diff
@@ -0,0 +1,139 @@
+    def makeString(s: String): BsonValue = new BsonString(s)
+    def makeNull: BsonValue = new BsonNull()
+    def makeBoolean(b: Boolean): BsonValue = new BsonBoolean(b)
+    def makeLocalDateTime(l: LocalDateTime) = errorImpossible
```
We recently added the rest of the temporal scalars to `ColumnType` in quasar, so they should be able to pass through for `ColumnType.Top` now.

cryogenian replied:
Wouldn't this break contract that `QDataEncode` and `QDataDecode` are reversing, since `QDataDecode`, `getLocalDateTime` and all other except `getOffsetDateTime` emit `errorImpossible`

wemrysi replied:
Yeah, if that's the scenario for `BSON` the :+1: , but why did you have to redefine it here then?

wemrysi replied:
Or did we just not have a `QDataEncode` for `BSON`?

cryogenian replied:
Right, we don't have it because why do we need it except tests? :)

cryogenian commented on datasource/src/main/scala/quasar/physical/mongo/interpreter/Cartesian.scala:
```diff
@@ -0,0 +1,85 @@
+
+        val instructions = is flatMap (_ flatMap {
+          case Aggregator.project(a @ E.Object(_*)) =>
+            List(Aggregator.project(defaultObject + a))
```
What do you mean? The sequence of action is following 

Having `Cartesian (Map { alias, (name, instruction) })`

+ Having cartesian for every cartouche we create a projection from projected key to alias, then we have `{alias: $name}`
+ Then we interpret all instrucitons on `alias` field
+ Then we remove `nonNull` aggregator and from this instructions
+ Then we make object `uniqueKey: { alias: interpretationResult}` 
+ Then we add `nonNull` back

wemrysi replied:
This appears to be modifying all cartouche projections, merging them with `defaultObject`, I was just confirming that this was correct for the projection in a cartouche like `Mask(. -> Object), Pivot(ExcludeId, Object), Project(.x.y.z)` since the project is not being applied to the input of the cartesian.

cryogenian replied:
Having `alias -> key -> Mask(. -> Object), Pivot(whatever, Object), Project(.x.y.z) ;; alias2 -> whatever` all constructed `$project`s need to have both `alias` and `alias2` fields, but most of interpretations constructs `$project` with only one key. So `$project` constructed for `alias` modified to have both keys by merging with default object, that is simply preservationi of all aliases in cartesian, i.e. in this example it would be `alias2 -> $alias2` for projects working with `alias` and vice versa `alias -> $alias` for projects working with `alias2`

wemrysi replied:
I see, thanks for explaining. :+1:

cryogenian commented on datasource/src/main/scala/quasar/physical/mongo/interpreter/Mask.scala:
```diff
@@ -0,0 +1,151 @@
+import shims._
+
+object Mask {
+  final case class TypeTree(types: Set[ColumnType], obj: Map[String, TypeTree], list: Map[Int, TypeTree])
```
This is for filtering, e.g. we have `foo.bar.baz: Int, foo.quux: String, foo[0]zoo: Object`. Instead of matching/filtering this creates an object expression that is set as a value of `$project` field and value of `foo` is set as either `null|undefined` or if (typeExpr foo) is object or array object or array with recursion results.

cryogenian commented on datasource/src/main/scala/quasar/physical/mongo/interpreter/Mask.scala:
```diff
@@ -0,0 +1,151 @@
+  private def parseTypeStrings(parseType: ColumnType): List[String] = parseType match {
+    case ColumnType.Boolean => List("bool")
+    case ColumnType.Null => List("null")
+    case ColumnType.Number => List("double", "long", "int", "decimal")
```
It doesn't work for older mongo, and this one works for all versions.

cryogenian commented on datasource/src/main/scala/quasar/physical/mongo/interpreter/Pivot.scala:
```diff
@@ -0,0 +1,90 @@
+  }
+
+  def toArray(key: String, projection: E.Projection, p: ColumnType.Vector): Aggregator = {
+    val refined: MongoExpression = p match {
```
Oh! That's cool!, I'll remove the check

rintcius commented:
@jsantos17 

> How about replacing the `QScript -> WorkflowBuilder` stage with a `ScalarStage -> WorkflowBuilder` stage instead of re-implementing everything from scratch? We can reuse most of the old connector that way, while doing partial pushdown. WDYT @rintcius?

I agree that it's good to have proper stages like WorkflowBuilder again, but I'm fine with moving there in a gradual way, on a by-need basis and based on the code in this PR. This would allow us to come up with something smaller and different than WorkflowBuilder, but of course we can still steal ideas from there. 

I have edited ch5394 with a summary of the most important follow-ups from this PR (feel free to expand and edit). Would approaching it like this work for you @jsantos17 ?

/cc @cryogenian

djspiewak commented:
@cryogenian Can you insert an emergency feature flag in here? Like check a system property or an envar or something? Something like `reform.mongo.pushdown.enabled=false`. If you get a `false`, don't evaluate any of the pushdown and just return the whole set of stages back for the rvalue interpreter. Default to enabled.

I just want to be able to bail out if something completely unforeseen happens on a customer's mongo cluster.

cryogenian commented:
@djspiewak 

> Can you insert an emergency feature flag in here? Like check a system property or an envar or something?

Sure thing! Do you think that envar is better than additional config field? Honestly I was thinking about making whole configuration for all pushdown in config (especially since we can edit it)

djspiewak commented:
@cryogenian For backend feature flags, I would prefer system properties. However, making it part of the config is even better. Again, I would default to on, but allowing users to toggle it off is great.

cryogenian commented:
@jsantos17 @djspiewak 
+ Added a couple of example and props spec. It's funny how complicated they are. Please note, that it's possible to squash `project`s in cartesian, I postponed this until optimization work done. Hopefully mongo will optimize these calls. 
+ Added `disablePushdown` field to config
+ Added `RenderTree` instances

cryogenian commented:
Oh! Also fixed `Mask` it should have `cond(isVectort(proj), vector, expr)` not `vector` for objects and arrays

cryogenian commented:
/cc @wemrysi

djspiewak commented:
@cryogenian Revision: can we generalize the `disablePushdown` to a `pushdownMode` with the following options: **Off**, **Conservative**, **Aggressive**. Whenever the frontend is connected to a backend which contains a version ≥4.x quasar-datasource-mongo, it should add this as a drop-down to the add datasource dialog for mongo. I think we can default to Aggressive, but I'm open to being convinced otherwise.

- **Off** No pushdown
- **Conservative** Pushdown for everything except `Cartesian`
- **Aggressive** Full pushdown

Is that possible? This will give users some more control over the load patterns.

jsantos17 commented on datasource/src/test/scala/quasar/physical/mongo/interpreter/CartesianSpec.scala:
```diff
@@ -0,0 +1,103 @@
+
+    val expectedDocs = expected map (_ map (_.toDocument))
+    val actualDocs = actual map (_ map (_.toDocument))
+    expectedDocs === actualDocs
```
Not asking for change, but it's sometimes nicer to use `beTreeEqual` from `quasar.matchers` to make these assertions

jsantos17 commented:
@alissapajer @wemrysi @nicflores @djspiewak @rintcius 

Should we attempt to run `it`s against this before merging? I'd like to have an idea on what works and what doesn't work.

djspiewak commented:
@jsantos17 Honestly, `ScalarStageSpec` is more comprehensive for this stuff than the integration tests are. I think it would be cool to be able to do what you're suggesting, but the infrastructure isn't in place at present, and I don't think it would find any new bugs, given how comprehensive the TCK is.

djspiewak commented:
Merging this now. Follow-up is ticketed!
